xdsl/backend/csl/lowering/convert_stencil_to_csl.py

```diff
@@ -0,0 +1,399 @@
+from attr import dataclass
+
+from xdsl.dialects import func
+from xdsl.dialects.arith import Cmpi, Constant, MinUI, Muli, OrI
+from xdsl.dialects.builtin import (
+    IntegerAttr,
+    IntegerType,
+    MemRefType,
+    ModuleOp,
+    Signedness,
+    StringAttr,
+    TensorType,
+)
+from xdsl.dialects.csl import (
+    ComptimeStructType,
+    ConcatStructOp,
+    ConstStructOp,
+    CslModuleOp,
+    GetColorOp,
+    ImportModuleConstOp,
+    LayoutOp,
+    MemberCallOp,
+    ModuleKind,
+    ModuleKindAttr,
+    ParamOp,
+    SetRectangleOp,
+    SetTileCodeOp,
+    SignednessCastOp,
+)
+from xdsl.dialects.scf import For, Yield
+from xdsl.dialects.stencil import ApplyOp
+from xdsl.ir import Attribute, Block, MLContext, Operation, Region
+from xdsl.passes import ModulePass
+from xdsl.utils.hints import isa
+from xdsl.utils.str_enum import StrEnum
+
+
+class Named(StrEnum):
+    """
+    Enum type to store names used across the program, for instance
+    names of parameters passed from the layout to the program module.
+    """
+
+    pattern = ("pattern",)
+    stencil_comms_params = ("stencil_comms_params",)
+    memcpy_params = ("memcpy_params",)
+    z_dim = ("z_dim",)
+    is_border_region_pe = ("is_border_region_pe",)
+
+
+class TranslationContext:
+    """
+    Helper class for translating to CSL.
+    Keeps track of stencil parameters such as stencil size, grid dimension.
+    Provides several methods to register any top-level params, imports, and consts/vars
+    both for program and layout modules. Params and imports are intrinsically names, consts/vars
+    should be given a name (which is only used for this translation)
+    """
+
+    grid_dim: tuple[int, ...] = (0, 0, 0)
+    pattern: int
+    program_module: CslModuleOp
+    layout_module: CslModuleOp
+    program_sym_name: StringAttr
+
+    program_params: dict[str, ParamOp] = {}
+    layout_params: dict[str, ParamOp] = {}
+    program_imports: dict[str, ImportModuleConstOp] = {}
+    layout_imports: dict[str, ImportModuleConstOp] = {}
+    program_vars: dict[str, Operation] = {}
+    layout_vars: dict[str, Operation] = {}
```

Comment:
Having mutable defaults here is a footgun of python, see e.g. https://docs.astral.sh/ruff/rules/mutable-dataclass-default/

You probably want to have these be initialized empty on every object instantiation.

Reply:
done as suggested